### ISSUE TITLE
inductor leaks memory

### ISSUE THREAD

**Issue by d4l3k** [pytorch/pytorch#96937]
### 🐛 Describe the bug

The inductor backend for `torch.compile` leaks memory on every call to a compiled mmseg model. This is currently causing my model to crash every ~9 hours because it eats up ~64gb of ram and crashes due to OOM.

Other backends such as aot_eager and cudagraphs show completely flat memory usage.

Repro:
https://github.com/d4l3k/torchdrive/blob/main/notebooks/inductor-memory-leak.ipynb

This showed up in both nightly from a month or so ago as well as in the latest release.

I don't have much idea on what the root cause is.

My repro is showing that it leaks ~22kb per call.

```
## compile_fn inductor
0 4604923904
100 4606889984
200 4609118208
300 4611346432
400 4613574656
500 4615802880
600 4618031104
700 4620259328
800 4622487552
900 4624715776

## compile_fn aot_eager
0 4785172480
100 4785172480
200 4785172480
300 4785172480
400 4785172480
500 4785172480
600 4785172480
700 4785172480
800 4785172480
900 4785172480

## compile_fn none
0 4914999296
100 4914999296
200 4914999296
300 4914999296
400 4914999296
500 4914999296
600 4914999296
700 4914999296
800 4914999296
900 4914999296
```

This seems to be unrelated to the AOTAutograd leak https://github.com/pytorch/pytorch/issues/94990 since it only occurs under inductor

Explicitly calling `gc.collect()` before measuring memory usage doesn't help with the leak.

### Versions

PyTorch version: 2.0.0+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Arch Linux (x86_64)
GCC version: (GCC) 12.2.1 20230201
Clang version: 15.0.7
CMake version: version 3.25.0
Libc version: glibc-2.37

Python version: 3.10.9 (main, Dec 19 2022, 17:35:49) [GCC 12.2.0] (64-bit runtime)
Python platform: Linux-6.2.2-arch1-1-x86_64-with-glibc2.37
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA GeForce RTX 2080
GPU 1: NVIDIA GeForce RTX 3070 Ti
GPU 2: NVIDIA GeForce RTX 3090

Nvidia driver version: 525.89.02
cuDNN version: Probably one of the following:
/opt/cudnn6/lib64/libcudnn.so.6.0.21
/usr/lib/libcudnn.so.8.6.0
/usr/lib/libcudnn_adv_infer.so.8.6.0
/usr/lib/libcudnn_adv_train.so.8.6.0
/usr/lib/libcudnn_cnn_infer.so.8.6.0
/usr/lib/libcudnn_cnn_train.so.8.6.0
/usr/lib/libcudnn_ops_infer.so.8.6.0
/usr/lib/libcudnn_ops_train.so.8.6.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   43 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          16
On-line CPU(s) list:             0-15
Vendor ID:                       AuthenticAMD
Model name:                      AMD Ryzen 7 3700X 8-Core Processor
CPU family:                      23
Model:                           113
Thread(s) per core:              2
Core(s) per socket:              8
Socket(s):                       1
Stepping:                        0
Frequency boost:                 enabled
CPU(s) scaling MHz:              90%
CPU max MHz:                     4426.1709
CPU min MHz:                     2200.0000
BogoMIPS:                        7189.95
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip rdpid overflow_recov succor smca sev sev_es
Virtualization:                  AMD-V
L1d cache:                       256 KiB (8 instances)
L1i cache:                       256 KiB (8 instances)
L2 cache:                        4 MiB (8 instances)
L3 cache:                        32 MiB (2 instances)
NUMA node(s):                    1
NUMA node0 CPU(s):               0-15
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Mitigation; untrained return thunk; SMT enabled with STIBP protection
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] mypy==0.971
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.24.1
[pip3] onnx2torch==1.5.4
[pip3] pytorch-memlab==0.2.4
[pip3] pytorch-msssim==0.2.1
[pip3] pytorch-triton==2.0.0+0d7e753227
[pip3] pytorch3d==0.7.2
[pip3] torch==2.0.0+cu118
[pip3] torch-tb-profiler==0.4.1
[pip3] torchaudio==2.0.1+cu118
[pip3] torchdrive==0.0.1
[pip3] torchmetrics==0.9.3
[pip3] torchtriton==2.0.0+0d7e753227
[pip3] torchvision==0.15.1+cu118
[pip3] triton==2.0.0

[pip3] sympy==1.11.1

cc @ezyang @gchanan @zou3519 @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

**Comment by d4l3k** [comment]
cc @bdhirsh @ezyang

**Comment by d4l3k** [comment]
I've found a much smaller repro -- can't say it's the only leak but this does cause a leak.

```py
def f_as_tensor(x):
    mean = torch.as_tensor((1,2,3,6,7,8,9,10), dtype=x.dtype, device=x.device)
    return mean
```

```
###############
## compile_fn inductor
0 635711488
10000 635973632
20000 636366848
30000 636760064
40000 637022208
50000 637284352
60000 637677568
70000 638070784
80000 638332928
90000 638726144
bytes/it 32.768
###############
## compile_fn cudagraphs
0 646852608
10000 646852608
20000 646852608
30000 646852608
40000 646852608
50000 646852608
60000 646852608
70000 646852608
80000 646852608
90000 646852608
bytes/it 0.0
###############
## compile_fn aot_eager
0 646852608
10000 646852608
20000 646852608
30000 646852608
40000 646852608
50000 646852608
60000 646852608
70000 646852608
80000 646852608
90000 646852608
bytes/it 0.0
###############
## compile_fn none
0 646852608
10000 646852608
20000 646852608
30000 646852608
40000 646852608
50000 646852608
60000 646852608
70000 646852608
80000 646852608
90000 646852608
bytes/it 0.0
```

**Comment by d4l3k** [comment]
Ran some more tests, leak seems to be consistent across most things

```
def f_as_tensor(x):
    """
    leaks 32bytes
    """
    mean = torch.as_tensor((1,2,3,6,7,8,9,10), dtype=x.dtype, device=x.device)
    return mean

def f_mul(x):
    """
    leaks 64bytes
    """
    return x * 10

def f_view(x):
    """
    no leak
    """
    return x.view(-1)
    
# leaks ~11kb/it
m = torchvision.models.resnet18()
"""

**Comment by d4l3k** [comment]
tracemalloc indicates that the difference is mostly in strings. 20MB in 2000 iterations of resnet18

```
warmup done
consumed 17432576
[ Top 10 ]
<string>:6: size=99.1 MiB (+21.7 MiB), count=3246694 (+712138), average=32 B
/home/rice/venvs/openape/lib/python3.10/site-packages/sympy/printing/printer.py:328: size=25.6 KiB (-48.6 KiB), count=408 (-783), average=64 B
/usr/lib/python3.10/tracemalloc.py:558: size=3136 B (-43.5 KiB), count=57 (-830), average=55 B
/home/rice/venvs/openape/lib/python3.10/site-packages/torch/autograd/function.py:506: size=16.6 KiB (+12.2 KiB), count=238 (+208), average=71 B
/home/rice/venvs/openape/lib/python3.10/site-packages/torch/_inductor/ir.py:4044: size=14.5 KiB (-6041 B), count=123 (-94), average=121 B
/home/rice/venvs/openape/lib/python3.10/site-packages/torch/_inductor/compile_fx.py:236: size=603 B (-1742 B), count=9 (-26), average=67 B
/tmp/ipykernel_3612779/3850902071.py:25: size=544 B (+544 B), count=1 (+1), average=544 B
/tmp/ipykernel_3612779/737906153.py:22: size=0 B (-544 B), count=0 (-1)
/tmp/ipykernel_3612779/3850902071.py:24: size=536 B (+536 B), count=1 (+1), average=536 B
/tmp/ipykernel_3612779/3850902071.py:15: size=0 B (-536 B), count=0 (-1)
```

Repro code:
```
import torch
from torchvision import models
import tracemalloc
import gc

device = torch.device('cuda')
m = models.resnet18().to(device)
m = torch.compile(m)
inp = torch.rand(2, 3, 240, 320, device=device)

# warmup
gc.collect()
tracemalloc.start()
for i in range(10):
    m(inp)
    
gc.collect()
    
print('warmup done')
snapshot1 = tracemalloc.take_snapshot()

start_mem = get_mem()
for i in range(2000):
    m(inp)
end_mem = get_mem()
print('consumed', end_mem-start_mem)

snapshot2 = tracemalloc.take_snapshot()
top_stats = snapshot2.compare_to(snapshot1, 'lineno')


print("[ Top 10 ]")
for stat in top_stats[:10]:
    print(stat)
````

**Comment by d4l3k** [comment]
String average size of 32B seems to line up with the 32 byte average for `as_tensor` and `64 bytes` for mul

**Comment by soumith** [comment]
@malfet can we get a patch release (and binaries) for Triton with this fix included. It seems like a pretty fundamental/nasty memory leak.
I presume we dont have to do a patch release of PyTorch for this.

**Comment by malfet** [comment]
Unassigning as fix has been landed to trunk, and 2.0.0 wheel has been published.

**Comment by d4l3k** [comment]
Uninstalling `triton` and reinstalling appears to have solved the issue. It's a bit annoying to have to manually muck around with that. Just rerunning pip install doesn't do anything so you either need to force install or manually uninstall triton first.

```
###############
## compile_fn inductor
0 2132410368
1000 2132410368
2000 2132410368
3000 2132410368
4000 2132410368
5000 2132410368
6000 2132410368
7000 2132410368
8000 2132410368
9000 2132410368
bytes/it 0.0
bytes 0
```
